web_poet/page_inputs.py

```diff
@@ -1,17 +1,187 @@
+        """
+        return cls([(pair["name"], pair["value"]) for pair in arg])
+
+    def declared_encoding(self) -> Optional[str]:
+        """ Return encoding detected from the Content-Type header, or None
+        if encoding is not found """
+        content_type = self.get('Content-Type', '')
+        return http_content_type_encoding(content_type)
+
+
+@attrs.define(auto_attribs=False, slots=False, eq=False)
+class HttpResponse:
+    """A container for the contents of a response, downloaded directly using an
+    HTTP client.
+
+    ``url`` should be a URL of the response (after all redirects),
+    not a URL of the request, if possible.
+
+    ``body`` contains the raw HTTP response body.
+
+    The following are optional since it would depend on the source of the
+    ``HttpResponse`` if these are available or not. For example, the responses
+    could simply come off from a local HTML file which doesn't contain ``headers``
+    and ``status``.
+
+    ``status`` should represent the int status code of the HTTP response.
+
+    ``headers`` should contain the HTTP response headers.
+
+    ``encoding`` encoding of the response. If None (default), encoding
+    is auto-detected from headers and body content.
     """
-    url: str
-    html: str
+
+    url: str = attrs.field()
+    body: HttpResponseBody = attrs.field(converter=HttpResponseBody)
+    status: Optional[int] = attrs.field(default=None)
+    headers: HttpResponseHeaders = attrs.field(factory=HttpResponseHeaders,
+                                               converter=HttpResponseHeaders)
+    _encoding: Optional[str] = attrs.field(default=None)
+
+    _DEFAULT_ENCODING = 'ascii'
+    _cached_text: Optional[str] = None
+
+    @property
+    def text(self) -> str:
+        """
+        Content of the HTTP body, converted to unicode
+        using the detected encoding of the response, according
+        to the web browser rules (respecting Content-Type header, etc.)
+        """
+        # Access self.encoding before self._cached_text, because
+        # there is a chance self._cached_text would be already populated
+        # while detecting the encoding
+        encoding = self.encoding
+        if self._cached_text is None:
+            fake_content_type_header = f'charset={encoding}'
+            encoding, text = html_to_unicode(fake_content_type_header, self.body)
+            self._cached_text = text
+        return self._cached_text
+
+    @property
+    def encoding(self):
+        """ Encoding of the response """
+        return (
+            self._encoding
+            or self._headers_declared_encoding()
+            or self._body_declared_encoding()
+            or self._body_inferred_encoding()
+        )
+
+    # XXX: see https://github.com/python/mypy/issues/1362
+    @property   # type: ignore
+    @memoizemethod_noargs
+    def selector(self) -> parsel.Selector:
+        # XXX: should we pass base_url=self.url, as Scrapy does?
```

Comment:
I'm afraid I don't quite follow this comment. Does it mean for a separate method like `urljoin` as opposed to being inside the `selector`?

Reply:
Yeah, sorry, it is unclear indeed. Scrapy does this: https://github.com/scrapy/scrapy/blob/9a28eb0bad1acf986d997905a410058f77911b7c/scrapy/selector/unified.py#L79 - setting base_url to response.url in Selector. That's one of the differences between Scrapy's Selector and parsel's Selector. But I'm not sure if that's the right thing to do (maybe it is).

Reply:
I see, thanks for explaining @kmike .

It might be worth adding the **base_url** with the notion of encouraging users that needs the relative lookups provided by the base_url. However, I'm not quite sure how common is this use case. Perhaps we can add this later on when user demand for it builds up.